pkg/controllers/state/cluster.go

```diff
@@ -262,13 +262,20 @@ func (c *Cluster) UpdateNode(ctx context.Context, node *v1.Node) error {
 	c.mu.Lock()
 	defer c.mu.Unlock()
 
+	managed := node.Labels[v1alpha5.ProvisionerNameLabelKey] != "" || node.Labels[v1beta1.NodePoolLabelKey] != ""
+	initialized := node.Labels[v1beta1.NodeInitializedLabelKey] != ""
 	if node.Spec.ProviderID == "" {
 		// If we know that we own this node, we shouldn't allow the providerID to be empty
-		if node.Labels[v1alpha5.ProvisionerNameLabelKey] != "" || (node.Labels[v1beta1.NodePoolLabelKey] != "") {
+		if managed {
 			return nil
 		}
 		node.Spec.ProviderID = node.Name
 	}
+	// If we have a managed node with no instance type label that hasn't been initialized,
+	// we need to wait until the instance type label gets propagated on it
+	if managed && node.Labels[v1.LabelInstanceTypeStable] == "" && !initialized {
```

Comment:
nit: any reason not to inline initialized since it doesn't look like you're using it otherwise?

Reply:
Also, looking at the code, doesn't this instance type label get propagated on machine registration ?

Reply:
It does. This is covering the case where we get a Node creation update for a node that we manage and it hasn't gotten its instance type label update yet from Karpenter.

This also is here for the machine migration case. We are about to be able to drop all of this logic so that we can just rely on the NodeClaim allocatable/capacity and not have to discover this information from the cloudprovider GetInstanceTypes()